File: TelerikUIExamples/TelerikUIExamples/ListViewLoadOnDemand.m
//
//  ListViewLoadOnDemand.m
//  TelerikUIExamples
//
//  Copyright (c) 2015 Telerik. All rights reserved.
//

#import <TelerikUI/TelerikUI.h>
#import "ListViewLoadOnDemand.h"
#import "CustomCardListViewCell.h"
#import "LoremIpsumGenerator.h"

@interface ListViewLoadOnDemand () <TKListViewDataSource, TKListViewDelegate>
@end

@implementation ListViewLoadOnDemand
{
    TKListView *_listView;
    TKDataSource *_names;
    TKDataSource *_photos;
    LoremIpsumGenerator *_loremIpsum;
    long _lastRetrievedDataIndex;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self addOption:@"Manual" selector:@selector(loadOnDemandManual) inSection:@"Load on demand mode"];
        [self addOption:@"Automatic" selector:@selector(loadOnDemandAuto) inSection:@"Load on demand mode"];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _lastRetrievedDataIndex = 15;
    _loremIpsum = [LoremIpsumGenerator new];
    _photos = [[TKDataSource alloc] initWithDataFromJSONResource:@"PhotosWithNames" ofType:@"json" rootItemKeyPath:@"photos"];
    _names = [[TKDataSource alloc] initWithDataFromJSONResource:@"PhotosWithNames" ofType:@"json" rootItemKeyPath:@"names"];
    
    _listView = [[TKListView alloc] initWithFrame:self.view.bounds];
    _listView.backgroundColor = [UIColor colorWithRed:0. green:1. blue:0. alpha:0.5];
    _listView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _listView.delegate = self;
    _listView.dataSource = self;
    
    // >> listview-buffer
    _listView.loadOnDemandBufferSize = 5;
    // << listview-buffer
    
    // >> listview-load-on-demand
    _listView.loadOnDemandMode = TKListViewLoadOnDemandModeManual;
    // << listview-load-on-demand
    
    _listView.contentInset = UIEdgeInsetsMake(10, 10, 10, 10);
    [self.view addSubview:_listView];
    [_listView registerClass:[CustomCardListViewCell class] forCellWithReuseIdentifier:@"cell"];
    
    TKListViewLinearLayout *layout = (TKListViewLinearLayout*)_listView.layout;
    layout.itemSize = CGSizeMake(100, 120);
    layout.itemSpacing = 5;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)loadOnDemandManual
{
    _listView.loadOnDemandMode = TKListViewLoadOnDemandModeManual;
    _lastRetrievedDataIndex = 15;
    [_listView reloadData];
    _listView.contentOffset = CGPointZero;
}

- (void)loadOnDemandAuto
{
    _listView.loadOnDemandMode = TKListViewLoadOnDemandModeAuto;
    _lastRetrievedDataIndex = 15;
    [_listView reloadData];
    _listView.contentOffset = CGPointZero;
}

#pragma mark - TKListViewDataSource

- (NSInteger)listView:(TKListView *)listView numberOfItemsInSection:(NSInteger)section
{
    return _lastRetrievedDataIndex;
}

// >> listview-load-on-demand-deque
- (TKListViewCell*)listView:(TKListView *)listView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    TKListViewCell *cell = [listView dequeueLoadOnDemandCellForIndexPath:indexPath];
    if (cell == nil) {
        cell = [listView dequeueReusableCellWithReuseIdentifier:@"cell" forIndexPath:indexPath];
        cell.imageView.image = [UIImage imageNamed:_photos.items[indexPath.row]];
        cell.textLabel.text = _names.items[indexPath.row];
        cell.detailTextLabel.text = [_loremIpsum randomString:10 + (arc4random()%16) forIndexPath:indexPath];
        cell.detailTextLabel.textColor = [UIColor whiteColor];
    }
    cell.backgroundView.backgroundColor = [UIColor colorWithWhite:0.3 alpha:0.5];
    ((TKView*)cell.backgroundView).stroke = nil;

    return cell;
}
// << listview-load-on-demand-deque

#pragma mark - TKListViewDelegate

// >> listview-should-load
- (BOOL)listView:(TKListView *)listView shouldLoadMoreDataAtIndexPath:(NSIndexPath *)indexPath
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        
        _lastRetrievedDataIndex = MIN(_names.items.count, _lastRetrievedDataIndex + 10);
        
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 2* NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            //Notifying the ListView that we have fresh data so it can hide the activity indicator and be ready for next load-on-demand request.
            
            if (_lastRetrievedDataIndex == _names.items.count) {
                listView.loadOnDemandMode = TKListViewLoadOnDemandModeNone;
            }
            
            [listView didLoadDataOnDemand];
        });
    });
    
    return YES;
}
// << listview-should-load

@end

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/ListViewLoadOnDemand.swift
//
//  ListViewLoadOnDemand.swift
//  TelerikUIExamplesInSwift
//
//  Copyright (c) 2015 Telerik. All rights reserved.
//

import UIKit

class ListViewLoadOnDemand: ExampleViewController, TKListViewDataSource, TKListViewDelegate {

    let listView = TKListView()
    let names = TKDataSource()
    let photos = TKDataSource()
    let loremIpsumGenerator = LoremIpsumGenerator()
    var lastRetrievedDataIndex = 15
    
    override init(nibName nibNameOrNil: String!, bundle nibBundleOrNil: NSBundle!) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        
        self.addOption("Manual", inSection:"Load on demand mode") { self.loadOnDemandManual() }
        self.addOption("Auto", inSection:"Load on demand mode") { self.loadOnDemandAuto() }
    }
    
    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        self.photos.loadDataFromJSONResource("PhotosWithNames", ofType: "json", rootItemKeyPath: "photos")
        self.names.loadDataFromJSONResource("PhotosWithNames", ofType: "json", rootItemKeyPath: "names")

        listView.frame = self.view.bounds
        listView.autoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight
        listView.backgroundColor = UIColor(red: 0.0, green: 1.0, blue: 0.0, alpha: 0.5)
        listView.dataSource = self
        listView.delegate = self
        listView.loadOnDemandBufferSize = 5
        listView.loadOnDemandMode = TKListViewLoadOnDemandMode.Manual
        listView.contentInset = UIEdgeInsetsMake(10, 10, 10, 10)
        self.view.addSubview(listView)
        listView.registerClass(CustomCardListViewCell.classForCoder(), forCellWithReuseIdentifier:"cell")
    
        let layout = listView.layout as! TKListViewLinearLayout
        layout.itemSize = CGSizeMake(100, 120)
        layout.itemSpacing = 5
    
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func loadOnDemandManual() {
        self.lastRetrievedDataIndex = 15
        self.listView.loadOnDemandMode = TKListViewLoadOnDemandMode.Manual
        self.listView.reloadData()
        self.listView.contentOffset = CGPoint.zeroPoint
    }

    
    func loadOnDemandAuto() {
        self.lastRetrievedDataIndex = 15
        self.listView.loadOnDemandMode = TKListViewLoadOnDemandMode.Auto
        self.listView.reloadData()
        self.listView.contentOffset = CGPoint.zeroPoint
    }
    
// MARK: TKListViewDataSource
    
    func listView(listView: TKListView!, numberOfItemsInSection section: Int) -> Int {
        return lastRetrievedDataIndex
    }
    
    func listView(listView: TKListView!, cellForItemAtIndexPath indexPath: NSIndexPath!) -> TKListViewCell! {
        
        var cell = listView.dequeueLoadOnDemandCellForIndexPath(indexPath)
        
        if (cell == nil) {
            cell = listView.dequeueReusableCellWithReuseIdentifier("cell", forIndexPath: indexPath) as! TKListViewCell
            cell.imageView.image = UIImage(named: self.photos.items[indexPath.row] as! String)
            cell.textLabel.text = names.items[indexPath.row] as? String
            cell.detailTextLabel.text = loremIpsumGenerator.randomString(10 + Int(arc4random_uniform(16)), indexPath: indexPath) as String
            cell.detailTextLabel.textColor = UIColor.whiteColor()
        }
        
        cell.backgroundView?.backgroundColor = UIColor(white: 0.3, alpha: 0.5)
        let backgroundView = cell.backgroundView as! TKView
        backgroundView.stroke = nil

        return cell
    }
    
// MARK: TKListViewDelegate
    
    func listView(listView: TKListView!, shouldLoadMoreDataAtIndexPath indexPath: NSIndexPath!) -> Bool {
       dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), { () -> Void in
        self.lastRetrievedDataIndex = min(self.names.items.count, self.lastRetrievedDataIndex + 10)
        
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, Int64(2 * Double(NSEC_PER_SEC))), dispatch_get_main_queue(), {
            //Notifying the ListView that we have fresh data so it can hide the activity indicator and be ready for next load-on-demand request.
            
            if self.lastRetrievedDataIndex == self.names.items.count {
                listView.loadOnDemandMode = TKListViewLoadOnDemandMode.None
            }
            
            listView.didLoadDataOnDemand()
        })
       })
        
        return true
    }
}
